src/Control/Exception/Annotated.hs

```diff
@@ -263,83 +280,115 @@ tryFlatten exn =
+-- As of 0.2.0.0, an alias for 'checkpointMany'.
 --
 -- @since 0.1.0.0
 checkpointCallStackWith
     :: (MonadCatch m, HasCallStack)
     => [Annotation]
     -> m a
     -> m a
-checkpointCallStackWith ann action =
-    action `Safe.catch` \(exn :: SomeException) ->
-        Safe.throw
-            . annotate ann
-            . addCallStackToException callStack
-            $ case Safe.fromException exn of
-                Just (e' :: AnnotatedException SomeException) ->
-                    case annotatedExceptionCallStack e' of
-                        Nothing ->
-                            annotate ann e'
-                        Just _preexistingCallstack ->
-                            e'
-                Nothing -> do
-                    annotate ann $ new exn
+checkpointCallStackWith anns =
+    withFrozenCallStack (checkpointMany anns)
 
--- | Add the current 'CallStack' to the checkpoint. This function searches any
--- thrown exception for a pre-existing 'CallStack' and will not overwrite or
--- replace the 'CallStack' if one is already present.
---
--- Primarily useful when you're wrapping a third party library.
+{-# DEPRECATED checkpointCallStackWith "As of 0.2.0.0 this is exactly equivalent to `checkpointMany`." #-}
+
+-- | Adds only the current 'CallStack' to the checkpoint. This function searches
+-- any thrown exception for a pre-existing 'CallStack' and will merge the given
+-- pre-existing 'CallStack' with the one on this function, in an attempt to
+-- preserve the actual call history.
 --
 -- @since 0.1.0.0
 checkpointCallStack
     :: (MonadCatch m, HasCallStack)
     => m a
     -> m a
 checkpointCallStack =
-    checkpointCallStackWith []
+    withFrozenCallStack (checkpoint (Annotation callStack))
 
 -- | Add the list of 'Annotation' to any exception thrown in the following
 -- action.
 --
 -- @since 0.1.0.0
-checkpointMany :: (MonadCatch m) => [Annotation] -> m a -> m a
-checkpointMany ann action =
+checkpointMany :: (MonadCatch m, HasCallStack) => [Annotation] -> m a -> m a
+checkpointMany anns action =
     action `Safe.catch` \(exn :: SomeException) ->
-        Safe.throw . annotate ann $ case Safe.fromException exn of
-            Just (e' :: AnnotatedException SomeException) ->
-                e'
-            Nothing -> do
-                new exn
+        Safe.throw
+            . addCallStackToException callStack
+            . annotate anns
+            $ case Safe.fromException exn of
+                Just (e' :: AnnotatedException SomeException) ->
+                    e'
+                Nothing -> do
+                    pure exn
 
 -- | Retrieves the 'CallStack' from an 'AnnotatedException' if one is present.
 --
+-- The library maintains an internal check that a single 'CallStack' is present
+-- in the list, so this only returns the first one found. If you have added
+-- a 'CallStack' directly to the @['Annotation']@, then this will likely break.
+--
```

Comment:
Okay, sorry for sounding like a broken record here, but this just corroborates my earlier thinking. Wouldn't it _really_ make more sense for `AnnotatedException` to have a `Maybe CallStack` field?